## Lab: SQL Practice

> Authors: Matt Brems & Jeff Hale

In this lab, you have 11 prompts. Most prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to help you practice your SQL - an **important** skill for working with data! 

Do everything with an SQL query in `pd.read_sql()`. 

If you want extra practice, do everything in pandas, too! 🐼

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The ERD for this database is here:

<img src="./images/erd.png" alt="erd" width="750"/>

In [1]:
# Imports

import pandas as pd
from sqlalchemy import create_engine

In [2]:
# Set up the engine to access the data.

# 'driver://username:password@host:port/database'
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.*
FROM stores as s;
"""

pd.read_sql(sql, engine)

store                         name store_status  \
0      2106  Hillstreet News and Tobacco            A   
1      2112               Mike's Liquors            I   
2      2113               Jamboree Foods            A   
3      2119           Manly Liquor Store            I   
4      2130         Sycamore Convenience            A   
...     ...                          ...          ...   
1968   9917       Honey Creek Distillery            A   
1969   9918  Paradise Distilling Company            A   
1970   9919            Dehner Distillery            A   
1971   9920        Louisiana Spirits LLC            A   
1972   9921                  maDIKwe USA            A   

                                          store_address address_info  
0     2217 College\nCedar Falls, IA 506130000\n(42.5...         None  
1     407 Sharp St.\nGlenwood, IA 515340000\n(41.046...         None  
2     1119 Market St. Box 71\nGowrie, IA 505430000\n...         None  
3     133 East Main\nManly, IA 504560000\n(43.286863...         None  
4     617 Sycamore\nWaterloo, IA 507030000\n(42.4978...         None  
...                                                 ...          ...  
1968  29538, 210th St.\nKeosauqua, IA 52565\n(40.740...         None  
1969  245, Railroad Ave\nDubuque, IA 52001\n(42.4873...         None  
1970  7500, University Ave\nClive, IA\n(41.600352803...         None  
1971  20909, South I-10 Frontage Rd\nLacassine, IA 7...         None  
1972  3101, Ingersoll Ave\nDes Moines, IA 50312\n(41...         None  

[1973 rows x 5 columns]

**Prompt 1 Answer**: 

1973 rows

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth? It's okay that numbers and capital letters come before lower case letter.

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.*
FROM stores as s
ORDER BY s.name;
"""

pd.read_sql(sql, engine)

store                                name store_status  \
0      4986  218 Fuel Express & Chubby's Liquor            A   
1      3047          3047 Cub Foods / Iowa City            I   
2      3060               3060 Cub Foods / Ames            I   
3      3061         3061 Cub Foods / Sioux City            I   
4      4845                  3rd St Convenience            A   
...     ...                                 ...          ...   
1968   3992       Xo Food And Liquor / Waterloo            I   
1969   4066                       Xpress Liquor            I   
1970   4886                      Yr Dollar Stop            A   
1971   4908                Zapf's Pronto Market            A   
1972   4874                      Z's Quickbreak            A   

                                          store_address address_info  
0     68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1     855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2     3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3     1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None  
4     216, 3rd St Se\nCedar Rapids, IA 52401\n(41.97...         None  
...                                                 ...          ...  
1968  428 Franklin St\nWaterloo, IA 507030000\n(42.5...         None  
1969  1201 North Jefferson Ste 100\nIndianola, IA 50...         None  
1970  1500 1st Ave Ne\nCedar Rapids, IA 52402\n(41.9...         None  
1971  107 1st St Sw\nElkader, IA 52043\n(42.85350191...         None  
1972  201, E Washington St\nMount Pleasant, IA 52641...         None  

[1973 rows x 5 columns]

**Prompt 2 Answer**:

3061 Cub Foods / Sioux City

### Prompt 3

For sales where the category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what are the bottle price and number of bottles in each of the first 3 transactions?

In [15]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT a.*
FROM sales as a
WHERE a.category_name LIKE 'IMPORTED VODKA'
ORDER BY a.vendor;
"""

pd.read_sql(sql, engine)

date convenience_store  store county_number   county category  \
0      2014-07-01              None   4509            85    Story  1032080   
1      2014-07-30              None   2548            77     Polk  1032080   
2      2014-06-19              None   4481            52  Johnson  1032080   
3      2014-09-29              None   3773            57     Linn  1032080   
4      2014-05-20              None   2619            77     Polk  1032080   
...           ...               ...    ...           ...      ...      ...   
124670 2014-02-06              None   2248            77     Polk  1032080   
124671 2014-02-17              None   4829            77     Polk  1032080   
124672 2014-02-05              None   2508            57     Linn  1032080   
124673 2014-02-19              None   2560            57     Linn  1032080   
124674 2014-02-27              None   2561            77     Polk  1032080   

         category_name vendor_no                     vendor    item  \
0       IMPORTED VODKA        14  Adamba Imports Int'l Inc.  904783   
1       IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   
2       IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   
3       IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   
4       IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   
...                ...       ...                        ...     ...   
124670  IMPORTED VODKA       239                       None   35109   
124671  IMPORTED VODKA       239                       None   35109   
124672  IMPORTED VODKA       239                       None   35109   
124673  IMPORTED VODKA       239                       None   35109   
124674  IMPORTED VODKA       239                       None   35109   

                            description  pack  liter_size state_btl_cost  \
0                          Vesica Vodka    12         750          $5.99   
1           Michael Godard "Mr. Smooth"     6         750         $19.00   
2           Michael Godard "Mr. Smooth"     6         750         $19.00   
3           Michael Godard "Mr. Smooth"     6         750         $19.00   
4           Michael Godard "Mr. Smooth"     6         750         $19.00   
...                                 ...   ...         ...            ...   
124670  Russian Standard Original Vodka    12         750         $10.89   
124671  Russian Standard Original Vodka    12         750         $10.89   
124672  Russian Standard Original Vodka    12         750         $10.89   
124673  Russian Standard Original Vodka    12         750         $10.89   
124674  Russian Standard Original Vodka    12         750         $10.89   

       btl_price  bottle_qty   total  
0          $8.99          12  107.88  
1         $28.50           6  171.00  
2         $28.50           3   85.50  
3         $28.50           3   85.50  
4         $28.50          30  855.00  
...          ...         ...     ...  
124670    $16.33          12  195.96  
124671    $16.33          12  195.96  
124672    $16.33          12  195.96  
124673    $16.33           3   48.99  
124674    $16.33          12  195.96  

[124675 rows x 17 columns]

**Prompt 3 Answer**: 


1. price: \$8.99, quantity: 12 bottles
2. price: \$19.00, quantity: 6 bottles
3. price: \$19.00, quantity: 3 bottles

### Prompt 4

What is the total amount of liquor sold in February 2015? Display it in dollars and cents.

In [26]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT CAST(SUM(a.total) as money)
FROM sales as a
WHERE DATE_PART('month', a.date) = 2 AND DATE_PART('year', a.date) = 2015;
"""
pd.read_sql(sql, engine)

sum
0  $21,295,350.55

**Prompt 4 Answer**: The total amount of liquor sold in February 2015 

\$21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? 

Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. 

Rename the columns to make sure there isn't any confusion as to what they mean.

In [35]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT a.vendor AS "Vendor", COUNT(a.vendor) AS "Transaction Count", SUM(a.bottle_qty) as "Bottles Sold", CAST(SUM(a.total) as money) as "Total Revenue"
FROM sales as a
WHERE a.category_name LIKE 'IMPORTED VODKA'
GROUP BY a.vendor
ORDER BY COUNT(a.vendor) DESC
LIMIT 10;
"""

pd.read_sql(sql, engine)

Vendor  Transaction Count  Bottles Sold  \
0  Pernod Ricard USA/Austin Nichols              42338        403450   
1               Bacardi U.S.A. Inc.              26441        236047   
2                   Diageo Americas              15725        152038   
3   Constellation Wine Company Inc.              12565        116053   
4                   Jim Beam Brands               5850        104119   
5                       Stoli Group               4443         35513   
6              Bacardi U.S.A., Inc.               3344         30239   
7              Imperial Brands Inc.               2799         21033   
8            Moet Hennessy USA Inc.               2153         10966   
9  Constellation Wine Company, Inc.               1672         12956   

   Total Revenue  
0  $7,880,365.88  
1  $5,771,242.07  
2  $3,346,425.42  
3  $1,895,005.51  
4  $1,807,097.23  
5    $832,071.19  
6    $765,347.31  
7    $265,317.79  
8    $323,126.54  
9    $200,283.64

**Prompt 5 Answer**: 

* Vendor: Pernod Ricard USA/Austin Nichols
* Total Transactions:42338
* Total Bottle Solder:403450
* Total Revenue: \$7,880,365.88

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of money charged for a product over the cost to make that product.

In Iowa, the law states the minimum price at which liquor may be sold.

Create a column called `Markup` that shows the difference between the `btl_price` and the `state_btl_cost`.

In [42]:
# Use a SQL query to display the table discussed above.

sql = """
SELECT a.btl_price, a.state_btl_cost, (a.btl_price - a.state_btl_cost) AS "Markup"
FROM sales as a;
"""

pd.read_sql(sql, engine)

btl_price state_btl_cost  Markup
0          $17.24         $11.49   $5.75
1          $17.24         $11.49   $5.75
2          $17.24         $11.49   $5.75
3          $25.73         $17.15   $8.58
4           $7.50          $5.00   $2.50
...           ...            ...     ...
3049908    $17.24         $11.49   $5.75
3049909    $29.99         $19.99  $10.00
3049910    $17.24         $11.49   $5.75
3049911    $17.24         $11.49   $5.75
3049912    $29.99         $19.99  $10.00

[3049913 rows x 3 columns]

### Prompt 7

Calculate and display the "Percentage Markup" per store (just store id, not name) by dividing markup by the state bottle cost. 

This will tell you how much stores are increasing their average price over the state minimum.

In [78]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT a.store, SUM(a.btl_price) as "Bottle Price", (SUM(a.btl_price) - SUM(a.state_btl_cost)) AS "Markup", ROUND((SUM(a.btl_price) - SUM(a.state_btl_cost))/(SUM(a.btl_price))*100) AS "Percentage Markup"
FROM sales as a
GROUP BY a.store;
"""

pd.read_sql(sql, engine)

store Bottle Price      Markup  Percentage Markup
0      2106  $112,082.81  $37,438.17               33.0
1      2113   $28,495.18   $9,513.08               33.0
2      2130   $81,331.17  $27,156.17               33.0
3      2152   $30,713.07  $10,302.83               34.0
4      2178   $55,591.58  $18,565.87               33.0
...     ...          ...         ...                ...
1347   9013       $13.64       $4.55               33.0
1348   9014       $45.42      $15.15               33.0
1349   9018      $291.00      $97.00               33.0
1350   9022       $19.01       $6.34               33.0
1351   9023       $26.39       $8.80               33.0

[1352 rows x 4 columns]

### Prompt 8

Calculate the average percentage markup per **store**. Return the average percentage markup and store number for the three stores with the largest markups.

In [126]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT a.store, SUM(a.btl_price) as "Bottle Price", (SUM(a.btl_price) - SUM(a.state_btl_cost)) AS "Markup", ROUND((SUM(a.btl_price) - SUM(a.state_btl_cost))/(SUM(a.btl_price))*100) AS "Percentage Markup"
FROM sales as a
GROUP BY a.store
ORDER BY "Percentage Markup" DESC
LIMIT 3;
"""

pd.read_sql(sql, engine)

store Bottle Price      Markup  Percentage Markup
0   2152   $30,713.07  $10,302.83               34.0
1   2510   $34,159.38  $11,456.18               34.0
2   2529   $68,294.77  $22,881.80               34.0

**Prompt 8 Answer**: The store with the highest average markup percentage is store 

1. Store #2152 - 34% Markup
2. Store #2510 - 34% Markup
3. Store #2529 - 34% Markup

### Prompt 9

Retrieve the three stores with the highest average markup percentage. Along with the store IDs and average markup percentage, return the minimum and maximum markup percentage for those stores, the names and addresses of those store.

What are the names of the 3 stores?

In [127]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT a.store as "Store ID", 
  MIN(s.name) AS "Name",
  MIN(s.store_address) AS "Address",
  ROUND((SUM(a.btl_price) - SUM(a.state_btl_cost))/(SUM(a.btl_price))*100) AS "Avg Percentage Markup",
  ROUND(MIN(((a.btl_price) - (a.state_btl_cost))/(a.btl_price)*100)) AS "Min Percentage Markup",
  ROUND(MAX(((a.btl_price) - (a.state_btl_cost))/((a.btl_price))*100)) AS "Max Percentage Markup"
FROM sales as a
INNER JOIN stores as s
  ON  a.store = s.store
GROUP BY a.store
ORDER BY "Avg Percentage Markup" DESC
LIMIT 3;
"""

pd.read_sql(sql, engine)

Store ID                                Name  \
0      2152                  Dugans Supermarket   
1      2510  Hy-Vee Drugstore #3 / Cedar Rapids   
2      2529  Hy-vee Drugstore #4 / Cedar Rapids   

                                             Address  Avg Percentage Markup  \
0  202 4th North\nRockwell, IA 504690000\n(42.986...                   34.0   
1  2405 Mt Vernon Rd Se\nCedar Rapids, IA 5240300...                   34.0   
2  4825 Johnson Ave Nw\nCedar Rapids, IA 52405000...                   34.0   

   Min Percentage Markup  Max Percentage Markup  
0                   33.0                   39.0  
1                   33.0                   41.0  
2                   33.0                   41.0

**Prompt 9 Answer**: The three stores with the highest markups are 

1. Dugans Supermarket
2. Hy-Vee Drugstore #3 / Cedar Rapids
3. Hy-Vee Drugstore #4 / Cedar Rapids

### Prompt 10

Which nine counties sell the highest number of bottles of liquor per capita? Calculate the average bottle size of the bottles of liquor sold by these counties, rounded to the nearest first decimal place. Also display the county name, and average bottle size.

Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

In [123]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT a.county, 
  SUM(a.bottle_qty)/MAX(c.population) AS "Bottles Sold per Capita",
  ROUND(AVG(a.liter_size),1) AS "Average Bottle Size"
FROM sales as a
  LEFT JOIN counties as c ON c.county = a.county
WHERE a.county NOT LIKE 'None'
GROUP BY a.county
ORDER BY "Bottles Sold per Capita" DESC
LIMIT 9;
"""

pd.read_sql(sql, engine)

county  Bottles Sold per Capita  Average Bottle Size
0      Dickinson                       22                974.2
1           Polk                       15                891.9
2     Black Hawk                       15                861.0
3    Cerro Gordo                       14                986.4
4          Scott                       13                869.9
5        Johnson                       13                895.2
6           Linn                       12                885.1
7  Pottawattamie                       11                890.9
8        Kossuth                       11               1040.5

**Prompt 10 Answer**: 

On average, Kossuth County sells the largest bottle size.

### Prompt 11

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions, if any, have a value of `total` that is not equal to `btl_price` times `bottle_qty`?

> You might get a type error. We cannot compare type `money` to type `real`. You may have to convert `total` to type `money` to compare the two directly.

In [ ]:
SELECT a.btl_price, a.bottle_qty, CAST(a.total AS money), a.btl_price*a.bottle_qty AS "Calculated Total", CAST(a.total AS money)-(a.btl_price*a.bottle_qty) AS "Difference"
FROM sales as a
CASE WHEN CAST(a.total AS money) = (a.btl_price*a.bottle_qty) THEN '0' ELSE '1' END AS DifferentAmount;

In [139]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT a.btl_price, a.bottle_qty, CAST(a.total AS money), a.btl_price*a.bottle_qty AS "Calculated Total", CAST(a.total AS money)-(a.btl_price*a.bottle_qty) AS "Difference"
FROM sales as a
ORDER BY "Difference" DESC;
"""


pd.read_sql(sql, engine)

btl_price  bottle_qty    total Calculated Total Difference
0          $13.74           6   $82.44           $82.44      $0.00
1          $14.63          24  $351.12          $351.12      $0.00
2          $20.79           2   $41.58           $41.58      $0.00
3           $2.89          24   $69.36           $69.36      $0.00
4          $29.99           6  $179.94          $179.94      $0.00
...           ...         ...      ...              ...        ...
3049908    $31.26           1   $31.26           $31.26      $0.00
3049909     $7.65           1    $7.65            $7.65      $0.00
3049910     $9.74           1    $9.74            $9.74      $0.00
3049911    $42.74           2   $85.48           $85.48      $0.00
3049912    $28.13           1   $28.13           $28.13      $0.00

[3049913 rows x 5 columns]

**Prompt 11 Answer**: 

I do not see any transaction where the total is not equal to the btl_price times bottle_qty.